libraries/standard/mqtt/src/mqtt.c

```diff
@@ -657,10 +1204,92 @@ MQTTStatus_t MQTT_Disconnect( MQTTContext_t * const pContext )
+                               uint32_t timeoutMs )
 {
-    return MQTTSuccess;
+    MQTTStatus_t status = MQTTBadParameter;
+    MQTTGetCurrentTimeFunc_t getTimeStamp = NULL;
+    uint32_t entryTime = 0U, remainingTimeMs = timeoutMs, elapsedTime = 0U;
+    MQTTPacketInfo_t incomingPacket;
+
+    if( pContext != NULL )
+    {
+        getTimeStamp = pContext->callbacks.getTime;
+        entryTime = getTimeStamp();
+        status = MQTTSuccess;
+    }
+    else
+    {
+        LogError( "MQTT Context cannot be NULL." );
+    }
+    
+    while( status == MQTTSuccess )
+    {
+        status = MQTT_GetIncomingPacketTypeAndLength( pContext->transportInterface.recv,
+                                                      pContext->transportInterface.networkContext,
+                                                      &incomingPacket );
+
+        if( status == MQTTNoDataAvailable )
+        {
+            /* Assign status so an error can be bubbled up to application,
+             * but reset it on success. */
+            status = handleKeepAlive( pContext );
+            if( status == MQTTSuccess )
+            {
+                /* Reset the status to indicate that we should not try to read
+                 * a packet from the transport interface. */
+                status = MQTTNoDataAvailable;
+            }
+        }
+        else if( status != MQTTSuccess )
+        {
+            LogErrorWithArgs( "Receiving incoming packet length failed. Status=%d",
+                              status );
+        }
+        else
+        {
+            /* Receive packet. Remaining time is recalculated at the end of the loop. */
+            status = receivePacket( pContext, incomingPacket, remainingTimeMs );
+        }
+
+        /* Handle received packet. If no data was read then this will not execute. */
+        if( status == MQTTSuccess )
+        {
+            incomingPacket.pRemainingData = pContext->networkBuffer.pBuffer;
+            /* PUBLISH packets allow flags in the lower four bits. For other
+             * packet types, they are reserved. */
+            if( ( incomingPacket.type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+            {
+                status = handleIncomingPublish( pContext, &incomingPacket );
+            }
+            else
+            {
+                status = handleIncomingAck( pContext, &incomingPacket );
+            }
+        }
+
+        if( status == MQTTNoDataAvailable )
+        {
+            /* No data available is not an error. Reset to MQTTSuccess so the
+             * return code will indicate success. */
+            status = MQTTSuccess;
+        }
+
+        if( status != MQTTSuccess )
+        {
+            LogErrorWithArgs( "Exiting receive loop. Error status=%d", status );
+        }
+
+        /* Recalculate remaining time and check if loop should exit. */
+        elapsedTime = calculateElapsedTime( getTimeStamp(), entryTime );
+        remainingTimeMs = timeoutMs - elapsedTime;
```

Comment:
This illustrates a couple of issues raised in my earlier comments – `calculateElapsedTime()` probably should not be a function since it just a subtraction; why aren't you using it on line 1285?  Also, `getTimeStamp()` should be named such that the units are obvious, e.g. `getTimeStampMs()`.

Reply:
Renamed to `getTimeStampMs()`. See above for reasoning on making `calculateElapsedTime()` its own function